llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5557,14 +5576,23 @@ InstructionCost LoopVectorizationCostModel::computePredInstDiscount(
 InstructionCost LoopVectorizationCostModel::expectedCost(ElementCount VF) {
   InstructionCost Cost;
 
+  // If the vector loop gets executed exactly once with the given VF, ignore the
+  // costs of comparison and induction instructions, as they'll get simplified
+  // away.
+  SmallPtrSet<Instruction *, 2> ValuesToIgnoreForVF;
+  auto TC = PSE.getSE()->getSmallConstantTripCount(TheLoop);
+  if (VF.isFixed() && TC == VF.getFixedValue() && !foldTailByMasking())
```

Comment:
If TC == VF, no tail should remain, so we shouldn't try to fold the tail by masking. Could you make this an assert instead?

Reply:
Of course you're right! There is code in computeMaxVF that disables tail-folding. In that case an assert does make sense.

Reply:
@fhahn Well, if I make it as assert, these tests fail:

```
  LLVM :: Transforms/LoopVectorize/AArch64/clamped-trip-count.ll
  LLVM :: Transforms/LoopVectorize/RISCV/low-trip-count.ll
  LLVM :: Transforms/LoopVectorize/RISCV/short-trip-count.ll
  LLVM :: Transforms/LoopVectorize/RISCV/truncate-to-minimal-bitwidth-cost.ll
```

Without assert two RISCV tests show different output (that's why I added this check in the first place)

Reply:
This is because in `computeMaxVF` the variable `MaxPowerOf2RuntimeVF` is bigger than the trip count and so the remainder of `MaxPowerOf2RuntimeVF/TC` is non-zero. Basically we only disable tail-folding if we can prove for every single possible VF there will not be a tail. So I believe @igogo-x86 is right and we need an explicit check.

Reply:
For example, suppose TC=8 and the max VF=16. In this case we *could* choose a VF of 16 and there will be a remainder, but the cost model may prefer VF=8 where there isn't a remainder.

Reply:
Thanks for checking, that makes sense.